### ISSUE TITLE
/dev/hidraw[x] unreliable and /dev/serial/by-id/bnot existant

### ISSUE THREAD

**Issue by DarkWarden85** [kukulich/home-assistant-jablotron100#71]
Hi,

I have the issue that my jablotron becomes unavailable at random times after a restart. This is because it switches between /dev/hidraw0 and /dev/hidraw1.
I would like to define it as /dev/serial/by-id/... in order to get rid of the problem. However, the jablotron usb-connection doesn't show up under serial, so I can't do that.
What it does is that it shows up under sda or sdb like so:

/dev/disk/by-id/usb-JABTRON_JA-100_Flexi_14004027628491-0:0

How do I have to define the port so that I get a reliable connection? Is there something wrong with my usb-connection?

Im running Home Assistant 2023.3.6 on HAOS on a raspberry Pi 4 with jablotron integration 3.17.

**Comment by kukulich** [comment]
I never have a problem with this type of switch on my Raspberry. It is sometimes reported when people use docker but not with HaOS on Rapsberry..

**Comment by DarkWarden85** [comment]
Wish I could report the same. It's an issue that happens randomly. Sometimes it's fine for a few months, other times it switches almost daily.

**Comment by enicky** [comment]
I have the same issue on an Intel NUC... It just switches the ports around

**Comment by kukulich** [comment]
You can try to ask on Discord, see README

**Comment by DarkWarden85** [comment]
It happened again yesterday after updating Home Assistant OS to version 10.0. I don't know why it happens so randomly. There are many restarts where the parts stay the same, other times though...

**Comment by kukulich** [comment]
There’s nothing I can do about it.

**Comment by fransward** [comment]
I have the same issue now also since the upgrade to 2023.4. Never had the problem before. Using HAOS on an Odroid-N2.
Changes daily between /dev/hidraw0 and /dev/hidraw1 making the device unavailable and thus failing some of the automations.

**Comment by standamo** [comment]
Maybe this would help:
```
# lsusb|grep -i jablotron
Bus 001 Device 006: ID 16d6:0008 Jablotron ltd JA-100 Flexi
# ll /sys/class/hidraw|grep -i :16d6:
lrwxrwxrwx    1 root     root             0 May 14 10:37 hidraw1 -> ../../devices/platform/scb/fd500000.pcie/pci0000:00/0000:00:00.0/0000:01:00.0/usb1/1-1/1-1.4/1-1.4:1.0/0003:16D6:0008.0002/hidraw/hidraw1
```

**Comment by PereViader** [comment]
It's also happening the same for me. Will research more

**Comment by enicky** [comment]
The thing I did right now is a nightly restart of Homa Assistant ... then it keeps on working for me

**Comment by DarkWarden85** [comment]
I think it would be great if we could specify some kind of fallback port for the integration to scan if the connection the other available port fails. Let's assume the integration usually connects to /dev/hidraw0 and we had a fallback port on /dev/hidraw1. If the integration now fails to connect on /dev/hidraw0, it tries to connect via /dev/hidraw1 once and only throws an error if both connections don't work. This would at least provide a workaround to the problem.

Mat beschte Gréiss/Cordialement

Michel Eischen
________________________________

**Comment by PereViader** [comment]
I don't really like the fallback approach, as it could be that someone has more stuff plugged in. But it would definitely work for me

**Comment by standamo** [comment]
I agree. Sending something to some device is risky. Why not lookup the right /dev/hidrawX based on vendor:device as I suggested above?

**Comment by DarkWarden85** [comment]
@standamo I find your suggestion difficult to understand. What do I enter into serial port field of the basic settings of the jablotron section based on your suggestion?

**Comment by PereViader** [comment]
Haven't tested, but this proposal looks possible and with no development from the integration side
https://chat.openai.com/share/c9b839a8-d582-459a-8a66-8228d8e12e11

**Comment by DarkWarden85** [comment]
> Haven't tested, but this proposal looks possible and with no development from the integration side https://chat.openai.com/share/c9b839a8-d582-459a-8a66-8228d8e12e11

Looks interesting, but be careful if openai isn't making stuff up in the details.

**Comment by standamo** [comment]
> @standamo I find your suggestion difficult to understand. What do I enter into serial port field of the basic settings of the jablotron integration based on your suggestion?

@DarkWarden85 Not suggesting fix for you to put into the field for hidraw, but suggesting that the integration should ideally lookup the correct port by itself by something like the above mentioned piece of code - i.e. look for /sys/class/hidraw/hidrawX which links to the expected vendor:device pair. And thus avoid the need for that field.

**Comment by filipek92** [comment]
I wrote autodetection feature in merge request #76. Testing in progress in my home instalation in proxmox virtual environment.

**Comment by DarkWarden85** [comment]
@filipek92 Wow that's nice. Can't wait for it :)

**Comment by kukulich** [comment]
Auto detection released in https://github.com/kukulich/home-assistant-jablotron100/releases/tag/3.20.0
